### PR TITLE
Sunset py2

### PR DESCRIPTION
**Pull request by megies**
### What does this PR do?

WIP Removes all Python 2 compatibility from the code base

### Why was it initiated?  Any relevant Issues?

Next major release after 1.2.0 will be Python 3 only.

**No PRs are to be merged into master before this one**

Supersedes #1613 

Covers most important action items in #2228, leaving open for now to track optional things to do

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [ ] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [ ] Any new or changed features have are fully documented.
- [ ] Significant changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

**Comment by megies** [comment]
WIP just drop a note here if you (intend to) work on this.

**Comment by ThomasLecocq** [comment]
ping I can do stuff if needed, just let me know.

**Comment by ThomasLecocq** [comment]
the Appveyor issue seems to be linked to this: https://github.com/pypa/setuptools/issues/2017

**Comment by megies** [comment]
Yeah this: https://github.com/pypa/setuptools/issues/65

I'm gonna stop here for a while. Most of the brute force regex patching is done, after adapting to that setuptools change it will likely be looking at failing tests in detail.

**Comment by megies** [comment]
@QuLogic I have a feeling you might use this mechanics to switch off libmseed for your packaging that has it packaged externally? So maybe you have some input on how to adapt to that setuptools change, perhaps?

**Comment by ThomasLecocq** [comment]
image
that won't help :-)

**Comment by megies** [comment]
you could try to debug it with `python -m pdb /path/to/obspy-runtests` maybe

**Comment by ThomasLecocq** [comment]
> you could try to debug it with `python -m pdb /path/to/obspy-runtests` maybe

it's not on my machine, it's on travisCI :-)

**Comment by megies** [comment]
Force pushed after getting the setuptools fix into master for 1.2.1

File: setup.py [hunk 1]
```diff
@@ -800,10 +786,7 @@ def setupPackage():
                 'Lesser General Public License (LGPL)',
             'Operating System :: OS Independent',
             'Programming Language :: Python',
-            'Programming Language :: Python :: 2',
-            'Programming Language :: Python :: 2.7',
             'Programming Language :: Python :: 3',
-            'Programming Language :: Python :: 3.4',
             'Programming Language :: Python :: 3.5',
```
Comment (hugovk) [inline setup.py]: I suggest also adding this param to `setup` to help pip know which version to install:

```python
python_requires=">=3.5",
```

Comment (megies) [reply setup.py]: Good idea

**Comment by ThomasLecocq** [comment]
@megies I'm jumping in now, first trying to get runstests report again

**Comment by ThomasLecocq** [comment]
ok, done.

This error: http://tests.obspy.org/109984/#2

is linked to the reading of  ``obspy\core\tests\data\example_py2.pickle`` 

which contains 

```   processingq]q (X~   ObsPy 1.0.3: trim(endtime=UTCDateTime(2009, 8, 24, 0, 20, 6)::fill_value=None::nearest_sample=True::pad=False::starttime=None)q!X�   ObsPy 1.0.3: trim(endtime=UTCDateTime(2009, 8, 24, 0, 20, 4, 130000)::fill_value=None::nearest_sample=True::pad=False::starttime=None)q"eX
   sampling_rateq#G@Y      X   stationq$X   RJOBq%X   locationq&X    q'X	   starttimeq(cobspy.core.utcdatetime
UTCDateTime
q))�q*}q+(U_UTCDateTime__precisionq,cfuture.types.newint
newint
q-��q.�q/}
```

So I suspect the unpickler is trying to set re-create the ``UTCDateTime`` object using the ``cfuture.types.newint`` object type, which miserably fails on a machine that has no ``future`` module installed.

**Comment by ThomasLecocq** [comment]
http://tests.obspy.org/109986/#3 is a reminder from the past: https://github.com/obspy/obspy/pull/2167

**Comment by megies** [comment]
Yeah, we won't be able to read old pickles on 2.0.0. We'll have to recreate test data for test cases involving pickle @ThomasLecocq

**Comment by d-chambers** [comment]
I just got a chance to take a look at the pickle failures now. One question: with so many other (better) supported options, do we really want to support/encourage pickle as a first class serialization method? Pickle has some serious drawbacks, including security and version/protocol compatibility issues. It can be great for quick, transitory things like sharing data between processes, but I don't think users should be turning to pickle for long-term storage.

So, for these tests, maybe we just pickle a stream to a `BytesIO` and ensure it can be re-inflated with the same `ObsPy` version rather than storing pickles in the data directory? I wouldn't mind dropping a note about not using pickle for long-term storage in the docs either.

**Comment by megies** [comment]
> So, for these tests, maybe we just pickle a stream to a `BytesIO` and ensure it can be re-inflated with the same `ObsPy` version rather than storing pickles in the data directory? I wouldn't mind dropping a note about not using pickle for long-term storage in the docs either.

Seems like a totally reasonable approach to me. I agree pickle should not be used if de/serialization is spanning Python or ObsPy versions.

**Comment by megies** [comment]
Alright bumped to Py 3.7 :rocket:

**Comment by ThomasLecocq** [comment]
@megies this line https://github.com/obspy/obspy/blob/33760901abb5998be6822d5ffe74db0b1f123968/.travis.yml#L70 should be called two times before the initial conda update conda.

**Comment by megies** [comment]
Sure thing, I'm not working on this right now in any case, just wanted to bump the Python version.

**Comment by megies** [comment]
@d-chambers you think you could take care of the pickle test change you proposed? :-)

**Comment by d-chambers** [comment]
Sure, I should have some time in a few hours. Any other significant items for this PR?

**Comment by megies** [comment]
There's this one test case that should be looked at: https://ci.appveyor.com/project/obspy/obspy/builds/31494678/job/69oqdo6uds5kb99o#L655

My feeling was it's pretty much redundant now, but somebody should really look at it.
Oh and we'll need to look at network tests: +TESTS:ALL

Also we'll bump to numpy 1.15 and maybe we should also check other dependency versions and then look through code to get rid of workarounds maybe. Also there's these action items in #2228

**Comment by d-chambers** [comment]
I made some progress. I need to work on a paper now but I should be able to come back to this next week. If anyone else wants to pick it up feel free.

There are a slew of network issues and potentially some matplotlib issues. I bumped the required MPL version to 3.2 as suggested by @heavelock  [here](https://github.com/obspy/obspy/issues/2228#issuecomment-601250940) and got a dozen or so failures locally but travis seems to have handled it better. We should probably take a closer look at the implications of upgrading matplotlib.

File: obspy/core/tests/test_utcdatetime.py [hunk 2]
```diff
@@ -1484,7 +1484,7 @@ def test_strftime_with_years_less_than_1900(self):
         t = UTCDateTime(1888, 1, 2, 1, 39, 37)
         self.assertEqual(t.strftime('%Y-%m-%d'), '1888-01-02')
         t = UTCDateTime(998, 11, 9, 1, 39, 37)
-        self.assertEqual(t.strftime('%Y-%m-%d'), '0998-11-09')
+        self.assertEqual(t.strftime('%Y-%m-%d'), '998-11-09')
```
Comment (megies) [inline obspy/core/tests/test_utcdatetime.py]: @d-chambers IIRC we had some code to make sure %Y is 4 characters always. It's long time ago so not sure

**Comment by megies** [comment]
We should probably deactivate all arclink tests and instead show a Deprecation Warning when arclink is used. The test server at GFZ is shut down and while we could switch to a different server for now, I don't think it's worth it.

**Comment by megies** [comment]
We should recreate all matplotlib test images on 3.2, maybe i'll get to it at some point

File: obspy/core/tests/test_utcdatetime.py (hunk 2 above)
Comment (d-chambers) [reply obspy/core/tests/test_utcdatetime.py]: Right you are, I had seen the python 3 check and assumed it wasn't needed any more. I re-added it and reverted the test.

**Comment by megies** [comment]
Good shout @d-chambers. Since this seems to be depending on external C implementations and not strictly depending on Python version or implementation, maybe an introspective approach might be safer. I've made a change, please have a look and let me know if you see a problem with it.

**Comment by d-chambers** [comment]
@megies,
The code is not particularly aesthetic, and I am not a huge fan of setting class attribute like that, but it certainly is a more comprehensive solution to a somewhat obscure edge case. A better way of doing it isn't coming to mind, so let's roll with it :wink:

**Comment by megies** [comment]
@ThomasLecocq can you run network tests on Windows for this one and send to report server? Also if somebody with Mac can run network tests for this one, that'd be awesome (@krischer?).

`-a` to run all tests including network tests
`--pr-url="https://github.com/obspy/obspy/pull/2577"` to make it easier to find the test when looking for it (it can be looked up by PR number then)

**Comment by ThomasLecocq** [comment]
running in an existing env (that has future etc)... will create a fresh empty one to test afterwards

**Comment by aringler-usgs** [comment]
I just ran it on a mac.  Let me know if you want this run again.

**Comment by megies** [comment]
Thanks @ThomasLecocq and @aringler-usgs. Looks like we have no systematic issues stemming from the Py2 compatibility layer drop. :rocket:

**Comment by ThomasLecocq** [comment]
@megies should I test in a fresh 3.7 ot 3.8 env ? do you have a conda create recipe?

**Comment by megies** [comment]
> @megies should I test in a fresh 3.7 ot 3.8 env ? do you have a conda create recipe?

I think you can just install 1.2.0 via packages in a fresh env and remove obspy and future, then install this branch with pip.

**Comment by megies** [comment]
I think this can be merged, but can never hurt for somebody else to give it another look. Other action items in #2228 seem secondary and can be worked on afterwards, I think, without conflicting with new PRs then too much.

**Comment by ThomasLecocq** [comment]
> > @megies should I test in a fresh 3.7 ot 3.8 env ? do you have a conda create recipe?
> 
> I think you can just install 1.2.0 via packages in a fresh env and remove obspy and future, then install this branch with pip.

so I did ``conda create -n TMP -c conda-forge python=3.7 obspy==1.2.0 --only-deps``  and removed ``future``

launching tests...

**Comment by ThomasLecocq** [comment]
done

**Comment by ThomasLecocq** [comment]
is there an easy way to get rid if those ``mopad`` warnings ? (I mean obviously rewriting the code is fine, but dunno if it's really needed?)

```
/home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/obspy/imaging/scripts/mopad.py:213: PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray.
  M[4], M[5], M[2]]).reshape(3, 3)
```

**Comment by megies** [comment]
@ThomasLecocq https://docs.python.org/3.7/library/warnings.html#warnings.filterwarnings if you feel like it

**Comment by heavelock** [comment]
I ran CI with GH Actions on my fork for Python 3.7 on all platforms and everything was green.
https://github.com/heavelock/obspy/pull/5

**Comment by megies** [comment]
OK, merging this now, I think most important steps have been taken and I'll leave #2228 open for now, since a few lower priority items are still open that can easily be handled later on in parallel to further (feature and other) work in master.

**Comment by megies** [comment]
:skull_and_crossbones: Python 2

**Comment by jkmacc-LANL** [comment]
🎉

**Comment by claudiodsf** [comment]
Python 2 RIP
